Gems/Terrain/Code/Source/Components/TerrainWorldDebuggerComponent.cpp

```diff
@@ -144,178 +158,252 @@ namespace Terrain
-                        AZ::Vector3(
-                            worldBounds.GetMin().GetX() + (xSector * xSectorSize), worldBounds.GetMin().GetY() + (ySector * ySectorSize),
-                            worldBounds.GetMin().GetZ()),
-                        AZ::Vector3(
-                            worldBounds.GetMin().GetX() + ((xSector + 1) * xSectorSize),
-                            worldBounds.GetMin().GetY() + ((ySector + 1) * ySectorSize), worldBounds.GetMax().GetZ()));
-
-                    sector.m_aabb.Clamp(worldBounds);
-
-                    m_wireframeSectors.push_back(AZStd::move(sector));
-                }
-            }
+        // To rebuild the wireframe, we walk through the sector by X, then by Y.  For each point, we add two lines in a _| shape.
+        // To do that, we'll need to cache the height from the previous point to draw the _ line, and from the previous row to draw
+        // the | line.
 
-            // Notify the visibility system that our bounds have changed.
-            AzFramework::IEntityBoundsUnionRequestBus::Broadcast(
-                &AzFramework::IEntityBoundsUnionRequestBus::Events::RefreshEntityLocalBoundsUnion, GetEntityId());
-        }
+        // When walking through the bounding box, the loops will be inclusive on one side, and exclusive on the other.  However, since
+        // our box is exactly aligned with grid points, we want to get the grid points on both sides in each direction, so we need to
+        // expand our query region by one extra point.
+        // For example, if our AABB is 2 m and our grid resolution is 1 m, we'll want to query (*--*--*--), not (*--*--).
+        // Since we're processing lines based on the grid points and going backwards, this will give us (*--*--*).
 
-        // For each sector, if it overlaps with the dirty region, clear it out and recache the wireframe line data.
-        for (auto& sector : m_wireframeSectors)
+        AZ::Aabb region = sector.m_aabb;
+        region.SetMax(region.GetMax() + AZ::Vector3(gridResolution.GetX(), gridResolution.GetY(), 0.0f));
+
+        // This keeps track of the height from the previous point for the _ line.
+        float previousHeight = 0.0f;
+
+        // This keeps track of the heights from the previous row for the | line.
+        AZStd::vector<float> rowHeights(aznumeric_cast<size_t>(ceil(region.GetExtents().GetX() / gridResolution.GetX())));
+
+        // We need 4 vertices for each grid point in our sector to hold the _| shape.
+        const uint32_t numSamplesX = static_cast<uint32_t>((region.GetMax().GetX() - region.GetMin().GetX()) / gridResolution.GetX());
+        const uint32_t numSamplesY = static_cast<uint32_t>((region.GetMax().GetY() - region.GetMin().GetY()) / gridResolution.GetY());
+        sector.m_lineVertices.clear();
+        sector.m_lineVertices.reserve(numSamplesX * numSamplesY * 4);
+
+        auto HeightCallback = [gridResolution, &previousHeight, &rowHeights,
```

Comment:
This function name seems a bit misleading - I was expecting a callback registered to some interface, but instead it looks like this just calculates sector vertices across the grid based on the previous / next heightfields, maybe something more descriptive like `PopulateSectorAtPosition`?

Reply:
In an upcoming change, the nested loops that are directly calling this lambda will get replaced with this:
```
        AzFramework::Terrain::TerrainDataRequestBus::Broadcast(
            &AzFramework::Terrain::TerrainDataRequests::ProcessHeightsFromRegion, region, gridResolution,
            AzFramework::Terrain::TerrainDataRequests::Sampler::EXACT,
            HeightCallback, nullptr
            );
```
I named it HeightCallback, because in that context, it's a callback that gets called from ProcessHeightsFromRegion with each height value in that region.  I'll rename it to ProcessHeightValue and add a comment above it, how's that sound?  It still might seem a bit odd right now, but hopefully it will look more natural once the ProcessHeightsFromRegion call is in place.

Reply:
Ah, yes, with that extra context it's completely reasonable (that's actually along the lines of what I was expecting, heh). 👍 for the name change though.